src/common/ad_tx.c

```diff
@@ -608,3 +612,397 @@ ad_tx_redo_act_next(struct ad_tx *tx)
+	if (tx->tx_layer != 0)
+		return 0;
+
+	rc = ad_tx_end(tx, err);
+	if (err == 0) {
+		if (rc == 0) {
+			ad_tx_stage_set(tx, UMEM_STAGE_ONCOMMIT);
+		} else {
+			D_ERROR("ad_tx_end(%d) failed, "DF_RC"\n", err, DP_RC(rc));
+			tx->tx_last_errno = rc;
+			ad_tx_stage_set(tx, UMEM_STAGE_ONABORT);
+		}
+	} else {
+		D_ASSERT(rc != 0);
+		D_ASSERTF(ad_tx_stage(tx) == UMEM_STAGE_ONABORT,
+			  "TX "DF_U64", bad stage %d\n", ad_tx_id(tx), ad_tx_stage(tx));
+	}
+
+	tx_callback(tx);
+	if (tx_get() == tx)
+		tx_set(NULL);
+	/* trigger UMEM_STAGE_NONE callback, this TX is finished but possibly with other WIP TX */
+	ad_tx_stage_set(tx, UMEM_STAGE_NONE);
+	tx_callback(tx);
+	rc = tx->tx_last_errno;
+	utx = ad_tx2umem_tx(tx);
+	D_FREE(utx);
+
+	return rc;
+}
+
+static int
+tx_abort(struct ad_tx *tx, int err)
+{
+	if (err == 0)
+		err = -DER_CANCELED;
+
+	ad_tx_stage_set(tx, UMEM_STAGE_ONABORT);
+
+	return tx_end(tx, err);
+}
+
+int
+mo_ad_tx_begin(struct umem_instance *umm, struct umem_tx_stage_data *txd)
+{
+	struct umem_tx		*utx = NULL;
+	struct ad_tx		*tx;
+	struct ad_blob_handle	 bh = umm2ad_blob_hdl(umm);
+	struct ad_blob		*blob;
+	struct umem_store	*store;
+	uint64_t		 tx_id;
+	int			 rc = 0;
+
+	blob = bh.bh_blob;
+	tx = tx_get();
+	if (tx == NULL) {
+		D_ALLOC_PTR(utx);
+		if (utx == NULL)
+			return -DER_NOMEM;
+
+		tx = umem_tx2ad_tx(utx);
+		rc = ad_tx_begin(bh, tx);
+		if (rc) {
+			D_ERROR("ad_tx_begin failed, "DF_RC"\n", DP_RC(rc));
+			D_FREE(utx);
+			return rc;
+		}
+
+		store = &blob->bb_store;
+		rc = store->stor_ops->so_wal_reserv(store, &tx_id);
+		if (rc) {
+			D_ERROR("so_wal_reserv failed, "DF_RC"\n", DP_RC(rc));
+			blob_decref(blob); /* drop ref taken in ad_tx_begin */
+			D_FREE(utx);
+			return rc;
+		}
+
+		/* possibly yield in so_wal_reserv, but tls_tx should be NULL when it get back */
+		D_ASSERT(tx_get() == NULL);
+		tx->tx_stage_cb = umem_stage_callback;
+		tx->tx_stage_cb_arg = txd;
+		ad_tx_id_set(tx, tx_id);
+		ad_tx_stage_set(tx, UMEM_STAGE_WORK);
+		tx_set(tx);
+		D_DEBUG(DB_TRACE, "TX "DF_U64" started\n", tx_id);
+	} else if (ad_tx_stage(tx) == UMEM_STAGE_WORK) {
+		if (blob != tx->tx_blob) {
+			D_ERROR("Nested TX for different blob\n");
+			rc = -DER_INVAL;
+			goto err_abort;
+		}
+		if (tx->tx_stage_cb_arg != txd) {
+			D_ERROR("Cannot set different TX callback argument\n");
+			rc = -DER_INVAL;
+			goto err_abort;
+		}
+		tx->tx_layer++;
+		D_DEBUG(DB_TRACE, "Nested TX "DF_U64", layer %d\n", ad_tx_id(tx), tx->tx_layer);
+	} else {
+		D_ERROR("Invalid stage %d to begin new tx\n", ad_tx_stage(tx));
+		rc = -DER_INVAL;
+		goto err_abort;
+	}
+
+	return 0;
+
+err_abort:
+	D_ASSERT(rc != 0);
+	D_ASSERT(tx != NULL);
+	rc = tx_abort(tx, rc);
+	return rc;
+}
+
+int
+mo_ad_tx_abort(struct umem_instance *umm, int err)
+{
+	struct ad_tx	*tx = tx_get();
+
+	D_ASSERTF(tx->tx_layer > 0,
+		  "TX "DF_U64", bad layer %d\n", ad_tx_id(tx), tx->tx_layer);
+	tx->tx_layer--;
```

Comment:
if we check if "tx->tx_layer == 0" in tx_commit(), then we probably should do the same at here?